### DISCUSSION TITLE
Append files through python by github actions

### DISCUSSION THREAD

**Discussion by Quantum-Codes** [community/community#23554] [New to GitHub]
<p><a href="https://github.com/Quantum-Codes/Testing-github" class="inline-onebox" rel="noopener nofollow ugc">GitHub - Quantum-Codes/Testing-github: Testing github actions</a> - this is the repo.</p>
<p>I am trying to append to an existing file <code>text.txt</code> with python using github actions.<br>
Currently, it does run the python file <code>code/main.py</code> but it doesn’t append to <code>text.txt</code><br>
(I created 2 text files: one in the code directory and one outside it. None of their content updates)</p>
<p>Can anyone say what’s wrong here?</p>

**Comment by jfullstackdev** [comment]
<p>maybe this is a similar situation as yours,</p>
<aside class="quote quote-modified" data-post="2" data-topic="17884">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/can-github-actions-directly-edit-files-in-a-repository/17884/2">Can GitHub actions directly edit files in a repository?</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    Hey <a class="mention" href="/jpadfield">@jpadfield</a>, you can most definitely have an action that edits internal files (except those in the “.github” folder.  
I do the following in one of my repo’s which calls a python script that parses all my source code and updates a todo.txt file and then commits it to the repo.  
jobs:
  build:

    runs-on: macos-latest

    steps:
    - uses: actions/checkout@v1
    - name: Run a one-line script
      run: python ./.github/workflows/todo_List.py
    - name: switching from HTTPS to SSH
      …
  </blockquote>
</aside>

**Comment by Quantum-Codes** [comment]
<p>So, directly using the <code>with open("file")</code> will not work and i must use a token?</p>

**Comment by jfullstackdev** [comment]
<p>even in GitHub API, editing a file requires authentication,</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Everything in the workspace is discarded after the workflow ends. If you want to keep changes, you need to commit and push them just like you would locally. The <code>actions/checkout</code> documentation <a href="https://github.com/actions/checkout#push-a-commit-using-the-built-in-token">has an example</a> of how to do that, usually the default <code>GITHUB_TOKEN</code> is enough for authentication.</p>

**Comment by Quantum-Codes** [comment]
<p>Thank you!! That example helped so much! Everything works great!<br>
Thanks to <a class="mention" href="/jdevstatic">@jdevstatic</a> and <a class="mention" href="/airtower-luna">@airtower-luna</a></p>
<p>The code:</p>
<pre><code class="lang-auto"># This workflow will install Python dependencies, run tests and lint with a single version of Python
# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions

name: Python application

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v2
    - name: Set up Python 3.10
      uses: actions/setup-python@v2
      with:
        python-version: "3.10"
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install flake8
        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
    - name: Lint with flake8
      run: |
        # stop the build if there are Python syntax errors or undefined names
        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
    - name: Run python
      run: |
        var=$(python code/main.py)
        echo $var
    - name: Push to github
      run: |
        git config user.name github-actions
        git config user.email github-actions@github.com
        git add .
        git commit -m "generated"
        git push
</code></pre>
